### ISSUE TITLE
Support ES `export * as ns from` form

### ISSUE THREAD

**Issue by mhegazy** [microsoft/TypeScript#4813]
The ES7 proposal is available at: https://github.com/leebyron/ecmascript-more-export-from

The additions include:

reexporting default:

``` ts
// proposed:
export v from "mod";

// symmetric to:
import v from "mod";
export {v};
```

reexporting as sub-module:

``` ts
// proposed:
export * as ns from "mod";

// symmetric to:
import * as ns from "mod";
export {ns};
```

also allowing combining 

``` ts
// proposed
export v, {x, y as w} from "mod"
// symmetric to
import v, {x, y as w} from "mod"
As well as

// proposed
export v, * as ns from "mod"
// symmetric to
import v, * as ns from "mod"
```

**Comment by vilic** [comment]
:+1:

**Comment by nevir** [comment]
This would be incredibly helpful for supporting intellisense for JavaScript projects in VS Code; `export * as foo from './foo';` seems pretty widely used

**Comment by aluanhaddad** [comment]
At first I thought this (the ECMAScript proposal) was a bad idea because the preponderance of existing export forms is a source of confusion for newcomers. However, I come around to the idea that the asymmetry between the import and export forms is actually a source of confusion in itself so adding this would be beneficial.

**Comment by jetpacmonkey** [comment]
Is there any chance of this being added sometime soon? I've been using this feature in my codebase with the `stage-1` babel preset, and was hoping to start using typescript to get gradually get some static typechecking, but typescript throws a bunch of errors every time it comes across `export x from './y';`

**Comment by mikew** [comment]
Our team's projects usually has a `common` file for developers to import common things from, without having to worry about where they come from:

```typescript
import {
  React,
  PropTypes,
  PureComponent,
  Button,
  connect,
} from './common'
```

Without `export from`, this `common` file will double in size and have needless duplication and some lines that can be confusing to newcomers.

```js
// This ...
export React, {
  PureComponent,
  PropTypes,
} from 'react'

// ... becomes ...
import * as _React from 'react'
export const React = _React
export const PropTypes = _React.PropTypes
export const PureComponent = _React.PureComponent
```

**Comment by mikew** [comment]
Even more strange that `export { connect } from 'react-redux'` works.

**Comment by nevir** [comment]
You can do:

export { default as React, PureComponent, PropTypes, } from 'react'

**Comment by mikew** [comment]
Almost, except then it's `module "@types/react/index" has no exported member 'default'`.

**Comment by aluanhaddad** [comment]
@mikew That is true but the same issue would apply the proposed

```ts
export React, {
  PureComponent,
  PropTypes,
} from 'react';
```
The `default` issue is a complex issue stemming from a convoluted set of incompatibilities and disparate assumptions between standards, languages, environments and transpilers.

However, it is entirely orthogonal to this proposal.

If you can write
```ts
import React from 'react';
```
and it works at _runtime_, set the `--allowSyntheticDefaultImports` flag.

**Comment by mjbvz** [comment]
#9906 tracked the proposed `export default from` ( https://github.com/tc39/proposal-export-default-from ) and was marked as a duplicate of this

Adding this comment to make sure the request for supporting `export default from` isn't potentially overlooked as part of this issue

**Comment by alexlur** [comment]
Is this on the roadmap or we have to wait until the proposal reaches a certain stage?

**Comment by kitsonk** [comment]
> Is this on the roadmap or we have to wait until the proposal reaches a certain stage?

IIRC, they tend to be taken on a case by case basis.  Pre Stage 3 tend to get implemented under a flag, and only if they are important enough.  Stage 3s tend to get implemented without flags.

**Comment by MitchStephan** [comment]
Is this still not supported? Or is there a setting I am missing in jsconfig?

**Comment by plantain-00** [comment]
> This spec has been withdrawn in favor of the equivalent micro-proposals: export ns from and export default from.

https://github.com/tc39/proposal-export-default-from
https://github.com/tc39/proposal-export-ns-from
They are both stage 1 now, usually, typescript will not support them until they are in stage 3.

**Comment by lukescott** [comment]
proposal-export-ns-from is being PR'd into the language spec https://github.com/tc39/ecma262/pull/1005, assuming consensus, which makes it official. Any plans to get this implemented into TypeScript?

**Comment by mhegazy** [comment]
Once the proposal makes it to stage 3, we will add support for it.

**Comment by lukescott** [comment]
OK, so if it's not listed in https://github.com/tc39/proposals at all (source of truth for stages), yet it's being PR'd into the language spec (https://github.com/tc39/ecma262/pull/1005) pending consensus (conceptually, but not technically stage 4), what stage is it right now?

**Comment by mhegazy** [comment]
@bterlson what stage is this proposal in at the moment?

**Comment by rclai** [comment]
I'd like to add that this form should also be supported if not already covered.

`export { default } from 'some-module'`

**Comment by mhegazy** [comment]
@rclai this should be already supported. ‘export default form “mod”’ is not.

**Comment by rclai** [comment]
I have a

``` 
// SomeComponent.js 
export default function SomeComponent() {
  return null
}

// index.js
export { default } from './SomeComponent'
```

When I import this from another place
```
import SomeComponent from 'components/SomeComponents
```
The autocomplete works, but Cmd clicking on the import doesn't take me to the `SomeComponent.js` and peeking doesn't work on it.

**Comment by rclai** [comment]
My `jsconfig.js`

```
{
  "compilerOptions": {
    "target": "es2015",
    "baseUrl": "./src",
    "paths": {
      "components/*": ["./components/**/*"]
    }
  }
}

```

**Comment by mhegazy** [comment]
Sorry my bad. I read it as ‘export { default as foo} from “mod”’.

Yes exporting the default is not supported yet

**Comment by rclai** [comment]
Okay cool, thanks.

**Comment by kitsonk** [comment]
@aMoniker the proposal was broken into two.  The `export * as ns from "mod"` is covered in a consensus PR at: https://github.com/tc39/ecma262/pull/1174 which still has not made it into the language specification.  Until that happens, TypeScript won't consider implementing it, though it is already shipping in most browsers.

The other part of the syntax, `export v from "mod"` is still a Stage 1 proposal: https://github.com/tc39/proposal-export-default-from and the TypeScript team won't consider implementing it until it reaches Stage 3 (or is added also via a consensus PR).

**Comment by weswigham** [comment]
`export * as ns from "foo"` was shipped in [v8 7.2](https://v8.dev/blog/v8-release-72), meaning current node and chrome can technically support it in their modules implementations - the downlevel isn't hard (just rewrite into 
```ts
import * as ns from "foo";
export { ns }
```
and feed down the transformer pipeline). This does imply a new module target that _doesn't_ downlevel it, though, which means a bump of the `exnext` module and the addition of `es2019` modules - the exact timing is a little muddy, since AFAIK it's shipped in v8 even though the proposal is still [stage 1](https://github.com/tc39/proposal-export-ns-from).

cc @RyanCavanaugh so we're aware

**Comment by Jessidhia** [comment]
`export * as ns from "foo"` is now officially part of the language https://github.com/tc39/ecma262/pull/1174

It probably should be preserved instead of downleveled in `esnext` / `es2020` emits.

---

Tangentially, this syntax is useful because of a bug in webpack 4 (fixed in webpack 5). It treats `export * as ns from ...` differently from the downlevel form, and is no longer capable of "tree shaking" if it's written in the downlevel form.

**Comment by DanielRosenwasser** [comment]
Changed the title since there is currently no export default version of this.

**Comment by Jessidhia** [comment]
It's still mis-titled because it's no longer "proposed"; it is _already_ in the spec 😉

**Comment by ExE-Boss** [comment]
I’ve&nbsp;now&nbsp;opened&nbsp;#35010 for&nbsp;the&nbsp;`export default from`&nbsp;form.

**Comment by DanielRosenwasser** [comment]
Thanks @kingwl!

**Comment by katis** [comment]
Previously I liked using TS namespaces and declaration merging for creating FP-style modules, without needing the `MyModule.T` convention:

```ts
export type Foobar = { }

export namespace Foobar {
  export function doStuff(foobar: Foobar) { }
}
```

```ts
import { Foobar } from './Foobar'

Foobar.doStuff()
```

Namespaces are not really the way to go any more, babel doesn't support them, they probably aren't tree-shakeable etc. But I was hoping that the `export * as ns` syntax could be used as a replacement, like this:

```ts
export type { Foobar } from './Foobar'

export * as Foobar from './Foobar'
```

This seems to pass the compiler, but the compiler only knows that Foobar is re-exported as a type, not a value. It would be great if the compiler could merge them into a combined value & type like with namespaces.

**Comment by rbuckton** [comment]
@katis: I filed #36792 with the content of your comment above. 

I'm closing *this* issue as it was fixed in #34903.
